pkg/capacityplanner/helpers.go

```diff
@@ -0,0 +1,341 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package capacityplanner
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
+	genV1 "github.com/dell/csi-baremetal/api/generated/v1"
+	acrcrd "github.com/dell/csi-baremetal/api/v1/acreservationcrd"
+	accrd "github.com/dell/csi-baremetal/api/v1/availablecapacitycrd"
+	"github.com/dell/csi-baremetal/pkg/base/k8s"
+	"github.com/dell/csi-baremetal/pkg/base/util"
+)
+
+// NewReservationHelper returns new instance of ReservationHelper
+func NewReservationHelper(logger *logrus.Entry, client *k8s.KubeClient,
+	capReader CapacityReader, resReader ReservationReader) *ReservationHelper {
+	return &ReservationHelper{
+		logger:    logger,
+		client:    client,
+		resReader: resReader,
+		capReader: capReader,
+	}
+}
+
+// ReservationHelper provides methods to create and release reservation
+type ReservationHelper struct {
+	logger *logrus.Entry
+	client *k8s.KubeClient
+
+	resReader ReservationReader
+	capReader CapacityReader
+
+	acMap       ACMap
+	acrMap      ACRMap
+	acNameToACR ACNameToACRNamesMap
+}
+
+// CreateReservation create reservation
+func (rh *ReservationHelper) CreateReservation(ctx context.Context, placingPlan *VolumesPlacingPlan) error {
+	logger := util.AddCommonFields(ctx, rh.logger, "ReservationHelper.CreateReservation")
+
+	volToAC := placingPlan.GetACsForVolumes()
+
+	var (
+		createErr   error
+		createdACRs = make([]*acrcrd.AvailableCapacityReservation, 0, len(volToAC))
+	)
+
+	for v, acs := range volToAC {
+		acsNames := make([]string, len(acs))
+		for i := 0; i < len(acs); i++ {
+			acsNames[i] = acs[i].Name
+		}
+		acrCR := rh.client.ConstructACRCR(genV1.AvailableCapacityReservation{
+			Name:         uuid.New().String(),
+			StorageClass: v.StorageClass,
+			Size:         v.Size,
+			Reservations: acsNames,
+		})
+		if createErr = rh.client.CreateCR(ctx, acrCR.Name, acrCR); createErr != nil {
+			createErr = fmt.Errorf("unable to create ACR CR %v for volume %v: %v", acrCR.Spec, v, createErr)
+			break
+		}
+		createdACRs = append(createdACRs, acrCR)
+	}
+	if createErr == nil {
+		return nil
+	}
+	// try to remove all created ACR
+	for _, acr := range createdACRs {
+		if err := rh.client.DeleteCR(ctx, acr); err != nil {
+			logger.Errorf("Unable to remove ACR %s: %v", acr.Name, err)
+		}
+	}
+	return createErr
+}
+
+// ReleaseReservation removes ACR for AC
+// if AC is in multiple ACRs, most suitable ACR will be remove, check choseACinACRWithLessAC function doc for details
+// Also, if AC was converted to AC with another SC, for example HDD-> HDDLVG,
+// we need to replace old AC with new in all ACRs
```

Comment:
As I understand one AC could be in a several ACRs **only** if AC storage class related to LVG. If `AC.StorageClass== HDD ` (or SSD, NVMe) extender must use this AC only in one ACR. And if it true we do not need logic from method `replaceACInACR`.

Reply:
Yes, we will call this method only if AC SC was change during volume creation
https://github.com/dell/csi-baremetal/blob/d84b8ad71c3e3680598695d481ed9813e6795a04/pkg/common/volume_operations.go#L147
https://github.com/dell/csi-baremetal/blob/d84b8ad71c3e3680598695d481ed9813e6795a04/pkg/capacityplanner/helpers.go#L117